src/graph/graphcontext.c

```diff
@@ -250,24 +250,35 @@ uint GraphContext_AttributeCount(GraphContext *gc) {
 }
 
 Attribute_ID GraphContext_FindOrAddAttribute(GraphContext *gc, const char *attribute) {
+	// If we are already in a critical region, no further locking should be done in this function.
+	bool in_critical_region = QueryCtx_IsLocked();
+
+	/* If we are not already in a critical region, acquire a read lock.
+	 * Reader threads may acquire a read lock multiple times as long
+	 * as they release the lock the same number of times.
+	 * Writer threads that have not yet entered the critical region will hold no lock at this point,
+	 * while writer threads within the critical region are guaranteed to be the only running thread. */
```

Comment:
I believe writer threads acquire a read lock in the same way read threads do, only difference is that writer threads will acquire write lock eventually (when performing the actual modification)